system/readline/readline_common.c

```diff
@@ -601,6 +601,8 @@ ssize_t readline_common(FAR struct rl_common_s *vtbl, FAR char *buf,
                           buf[nch++] = g_cmdhist.buf[idx][i];
                           RL_PUTC(vtbl, g_cmdhist.buf[idx][i]);
                         }
+
+                      buf[nch] = '\0';
```

Comment:
This should probably be just before the strcmp, otherwise if you run `foobar` and `fooba`, then the latter doesn't get saved.